apiserver/Makefile

```diff
@@ -19,27 +42,16 @@ jupyter_notebook_config.py:
 	    -e "s,@keyfile@,$(GSA_KEY_FILE),g" \
             < jupyter_notebook_config.py.in > $@
 
-build: install-hail-locally jupyter_notebook_config.py
-	docker build -t spark-base -f Dockerfile.spark-base .
-	docker build -t hail-base -f Dockerfile.hail-base .
-	docker build -t spark-master -f Dockerfile.master .
-	docker build -t spark-worker -f Dockerfile.worker .
-	docker build -t apiserver -f Dockerfile.apiserver .
-	docker build -t hail-jupyter -f Dockerfile.hail-jupyter .
-
-push: SPARK_MASTER_IMAGE="gcr.io/$(PROJECT)/spark-master:$(shell docker images -q --no-trunc spark-master | sed -e 's,[^:]*:,,')"
-push: SPARK_WORKER_IMAGE="gcr.io/$(PROJECT)/spark-worker:$(shell docker images -q --no-trunc spark-worker | sed -e 's,[^:]*:,,')"
-push: APISERVER_IMAGE="gcr.io/$(PROJECT)/apiserver:$(shell docker images -q --no-trunc apiserver | sed -e 's,[^:]*:,,')"
-push: HAIL_JUPYTER_IMAGE="gcr.io/$(PROJECT)/hail-jupyter:$(shell docker images -q --no-trunc hail-jupyter | sed -e 's,[^:]*:,,')"
-push: build
-	docker tag spark-master $(SPARK_MASTER_IMAGE)
-	docker push $(SPARK_MASTER_IMAGE)
-	docker tag spark-worker $(SPARK_WORKER_IMAGE)
-	docker push $(SPARK_WORKER_IMAGE)
-	docker tag apiserver $(APISERVER_IMAGE)
-	docker push $(APISERVER_IMAGE)
-	docker tag hail-jupyter $(HAIL_JUPYTER_IMAGE)
-	docker push $(HAIL_JUPYTER_IMAGE)
+$(eval $(call DOCKER_IMAGE,spark-base,install-hail-locally jupyter_notebook_config.py,SPARK_BASE_IMAGE))
+$(eval $(call DOCKER_IMAGE,hail-base,install-hail-locally jupyter_notebook_config.py,HAIL_BASE_IMAGE))
+$(eval $(call DOCKER_IMAGE,spark-master,install-hail-locally jupyter_notebook_config.py,SPARK_MASTER_IMAGE))
+$(eval $(call DOCKER_IMAGE,spark-worker,install-hail-locally jupyter_notebook_config.py,SPARK_WORKER_IMAGE))
+$(eval $(call DOCKER_IMAGE,apiserver,install-hail-locally jupyter_notebook_config.py,APISERVER_IMAGE))
+$(eval $(call DOCKER_IMAGE,hail-jupyter,install-hail-locally jupyter_notebook_config.py,HAIL_JUPYTER_IMAGE))
+
+build: build-spark-base build-hail-base build-spark-master build-spark-worker build-apiserver build-hail-jupyter
+
+push: push-spark-base push-hail-base push-spark-master push-spark-worker push-apiserver push-hail-jupyter
```

Comment:
hail-base was previously built but not pushed. Are there caching benefits from pushing?

Same for spark-base.

Reply:
Yeah, the `DOCKER_IMAGE` macro pulls the `latest` tagged image and uses it as a cache source. Without a push to the `latest` tag, there is no cache. Pushing the hash-tagged image is unnecessary for cache'ing but splitting push into `push-latest` and `push-hash` is annoying and doesn't change performance much because pushing an image twice with two different names only transmits the image once.

Reply:
Thanks. Ah, so if the latest tagged image is identical to a locally built copy, it will skip rebuild?